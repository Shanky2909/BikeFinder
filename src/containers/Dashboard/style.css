body{
    background-image: url('../../assests/background-image.png');
}
.pagination {
    display: flex;
    padding: 20px;
    justify-content: center;
}

.page-count {
    background-color: rgb(212, 242, 185);
    border-radius: 5px;
    padding: 10px;
}

.loader{
    position: relative;
    left: 49%;
    top: 49%;
    color: rgb(240, 41, 6);
}