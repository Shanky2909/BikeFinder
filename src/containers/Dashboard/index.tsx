import React, { Fragment, useEffect, useState } from 'react';
import axios from 'axios';
import { BikeApiParams } from '../../../public/interfaces';
import GlobalHeader from '../../components/GlobalHeader';
import CircularProgress from '@material-ui/core/CircularProgress';
import Cards from '../../components/Cards';
import { Pagination } from '@material-ui/lab';
import usePagination from '../../components/Pagination';
import './style.css';

const Dashboard: React.FC = () => {
  const [bikeData, setBikeData] = useState<Array<Object>>([]);
  const [pageNumber, setPageNumber] = useState<number>(1);
  const [loading, setLoading] = useState<boolean>(false);
  const [pageCount, setPageCount] = useState<number>(0);
  const [totalStolenCount, setTotalStolenCount] = useState<number>(0);
  const [searchQuery,setSearchQuery] = useState<string>('');
  const _data = usePagination(bikeData, 10);

  // Handleing Page Change
  const handleChange = (e: any, p: number) => {
    setPageNumber(p);
    _data.jump(p);
  };

  //Handleing Search
  const handleSearch = (value:string)=>{
    setSearchQuery(value);
  }

  //Getting Bike Data from API
  React.useEffect(() => {
    async function getBikeData(apiParams: BikeApiParams) {
      setLoading(true);
      axios({
        method: 'get',
        url: `https://bikeindex.org:443/api/v3/search`,
        withCredentials: false,
        params: apiParams,
      })
        .then((responseData: any) => {
          if (responseData) {
            const { data } = responseData;
            setBikeData(data.bikes);
          }
          setLoading(false);
        })
        .catch((error) => {
          console.log(error);
          setLoading(false);
        });
    }
    const apiParams: BikeApiParams = {
      page: pageNumber,
      per_page: 10,
      query:searchQuery,
      location: 'Sydney',
      stolenness: 'stolen',
      access_token: '2b8Qd-E0WdW0Q1FAhKFxH3VAtod7mGbOVBZ82nyTwWk',
    };
    getBikeData(apiParams);
  }, [pageNumber,searchQuery]);

  // Getting Total Page Count of Bikes stolen in Sydney 
  useEffect(() => {
    async function getPAginationCount(apiParams: BikeApiParams) {
      setLoading(true);
      axios({
        method: 'get',
        url: `https://bikeindex.org:443/api/v3/search/count`,
        withCredentials: false,
        params: apiParams,
      })
        .then((responseData: any) => {
          if (responseData) {
            const { data } = responseData;
            const { stolen } = data;
            let pages = parseInt(stolen) / 10 + '';
            setPageCount(parseInt(pages));
            setTotalStolenCount(stolen);
          }
          setLoading(false);
        })
        .catch((error) => {
          console.log(error);
          setLoading(false);
        });
    }
    const apiParams: BikeApiParams = {
      location: 'Sydney',
      stolenness: 'stolen',
      access_token: '2b8Qd-E0WdW0Q1FAhKFxH3VAtod7mGbOVBZ82nyTwWk',
    };
    getPAginationCount(apiParams);
  }, []);

  return (
    <Fragment>
      <GlobalHeader count={totalStolenCount} handleSearch={handleSearch}/>
      <div className='pagination'>
        <Pagination
          count={pageCount}
          size='large'
          page={pageNumber}
          variant='outlined'
          shape='rounded'
          className='page-count'
          onChange={handleChange}
        />
      </div>
      {loading ? (
        <CircularProgress color='inherit' className='loader'/>
      ) : (
        <Cards bikeData={bikeData} />
      )}
      <div className='pagination'>
        <Pagination
          count={pageCount}
          size='large'
          page={pageNumber}
          variant='outlined'
          shape='rounded'
          onChange={handleChange}
          className='page-count'
        />
      </div>
    </Fragment>
  );
};

export default Dashboard;

// //[7:03 pm] Ranjan Mishra
// https://github.com/SHOOTSTA/coding-challange
// ​[7:03 pm] Ranjan Mishra
//     https://bikeindex.org/documentation/api_v3#!/search/GET_version_search_format_get_0
// API v3The best bike registry: Simple, secure and free.bikeindex.org​[7:04 pm] Ranjan Mishra
//     2b8Qd-E0WdW0Q1FAhKFxH3VAtod7mGbOVBZ82nyTwWk
// ​[7:04 pm] Ranjan Mishra
//     https://bikeindex.org/bikes?serial=&button=&location=Sydney&distance=300&stolenness=proximity
// Search resultsSearch for bikes that have been registered on Bike Indexbikeindex.org
