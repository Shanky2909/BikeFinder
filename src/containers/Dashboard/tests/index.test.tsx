import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import Dashboard from '..';

jest.mock('axios');
const axios = require('axios');


describe('Dashboard', () => {
  test('renders learn Dashboard', async () => {
    axios.get = jest.fn().mockImplementation(() =>
      Promise.resolve({
        data: [
          {
            date_stolen: 1666102625,
            description: null,
            frame_colors: ['Black'],
            frame_model: 'Element C50',
            id: 1419598,
            is_stock_img: false,
            large_img:
              'https://files.bikeindex.org/uploads/Pu/649347/large_DE763928-31CC-4489-A27F-329547DF62A9.jpeg',
            location_found: null,
            manufacturer_name: 'Rocky Mountain Bicycles',
            external_id: null,
            registry_name: null,
            registry_url: null,
            serial: 'SRMVCTHEE9S3903',
            status: 'stolen',
            stolen: true,
            stolen_coordinates: [51.02, -114.11],
            stolen_location: 'Calgary, T2T 4V8, CA',
            thumb:
              'https://files.bikeindex.org/uploads/Pu/649347/small_DE763928-31CC-4489-A27F-329547DF62A9.jpeg',
            title: '2022 Rocky Mountain Bicycles Element C50',
            url: 'https://bikeindex.org/bikes/1419598',
            year: 2022,
          },
        ],
        status: 200,
        statusText: 'Ok',
        headers: {},
        config: {},
      })
    );
    axios.get = jest.fn().mockImplementation(() =>
      Promise.resolve({
        data: [
          {
            non: 807664,
            proximity: 33,
            stolen: 121642,
          },
        ],
        status: 200,
        statusText: 'Ok',
        headers: {},
        config: {},
      })
    );
    const wrapper = renderer.create(<Dashboard />);
    expect(wrapper.toJSON()).toMatchSnapshot()
  });
});
