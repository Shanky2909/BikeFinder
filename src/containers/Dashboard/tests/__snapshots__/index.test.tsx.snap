// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders learn Dashboard 1`] = `
[
  <div
    className="header"
  >
    <div
      className="heading"
    >
      <p>
        Bike Finder
      </p>
    </div>
    <div
      className="search-input"
    >
      <input
        className="input-search"
        onChange={[Function]}
        placeholder="Seach Here!"
        type="search"
      />
      <button
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </button>
    </div>
    <div
      className="text-content"
    >
      <p>
        Total Count: 
        0
      </p>
    </div>
  </div>,
  <div
    className="pagination"
  >
    <nav
      aria-label="pagination navigation"
      className="MuiPagination-root page-count"
    >
      <ul
        className="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>,
  <div
    className="pagination"
  >
    <nav
      aria-label="pagination navigation"
      className="MuiPagination-root page-count"
    >
      <ul
        className="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-outlined MuiPaginationItem-rounded Mui-disabled MuiPaginationItem-sizeLarge Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>,
]
`;
